tests/flytekit/unit/core/test_serialization.py

```diff
@@ -475,3 +486,302 @@ def z(a: int, b: str) -> typing.Tuple[int, str]:
+
+    assert wf_no_input() == default_val
+    assert wf_with_input() == input_val
+    assert wf_with_sub_wf() == (default_val, input_val)
+
+
+def test_default_args_task_optional_type_default_none():
+    default_val = None
+    input_val = 100
+
+    @task
+    def t1(a: typing.Optional[int] = default_val) -> typing.Optional[int]:
+        return a
+
+    @workflow
+    def wf_no_input() -> typing.Optional[int]:
+        return t1()
+
+    @workflow
+    def wf_with_input() -> typing.Optional[int]:
+        return t1(a=input_val)
+
+    @workflow
+    def wf_with_sub_wf() -> tuple[typing.Optional[int], typing.Optional[int]]:
+        return (wf_no_input(), wf_with_input())
+
+    wf_no_input_spec = get_serializable(OrderedDict(), serialization_settings, wf_no_input)
+    wf_with_input_spec = get_serializable(OrderedDict(), serialization_settings, wf_with_input)
+
+    assert wf_no_input_spec.template.nodes[0].inputs[0].binding.value == Scalar(
+        union=Union(
+            value=Literal(
+                scalar=Scalar(
+                    none_type=Void(),
+                ),
+            ),
+            stored_type=LiteralType(
+                simple=SimpleType.NONE,
+                structure=TypeStructure(tag="none"),
+            ),
+        ),
+    )
+    assert wf_with_input_spec.template.nodes[0].inputs[0].binding.value == Scalar(
+        primitive=Primitive(integer=input_val)
+    )
+
+    output_type = LiteralType(
+        union_type=UnionType(
+            [
+                LiteralType(
+                    simple=SimpleType.INTEGER,
+                    structure=TypeStructure(tag="int"),
+                ),
+                LiteralType(
+                    simple=SimpleType.NONE,
+                    structure=TypeStructure(tag="none"),
+                ),
+            ]
+        )
+    )
+    assert wf_no_input_spec.template.interface.outputs["o0"].type == output_type
+    assert wf_with_input_spec.template.interface.outputs["o0"].type == output_type
+
+    assert wf_no_input() == default_val
+    assert wf_with_input() == input_val
+    assert wf_with_sub_wf() == (default_val, input_val)
+
+
+def test_default_args_task_optional_type_default_int():
+    default_val = 10
+    input_val = 100
+
+    @task
+    def t1(a: typing.Optional[int] = default_val) -> typing.Optional[int]:
+        return a
+
+    @workflow
+    def wf_no_input() -> typing.Optional[int]:
+        return t1()
+
+    @workflow
+    def wf_with_input() -> typing.Optional[int]:
+        return t1(a=input_val)
+
+    @workflow
+    def wf_with_sub_wf() -> tuple[typing.Optional[int], typing.Optional[int]]:
+        return (wf_no_input(), wf_with_input())
+
+    wf_no_input_spec = get_serializable(OrderedDict(), serialization_settings, wf_no_input)
+    wf_with_input_spec = get_serializable(OrderedDict(), serialization_settings, wf_with_input)
+
+    assert wf_no_input_spec.template.nodes[0].inputs[0].binding.value == Scalar(
```

Comment:
actually sorry, i think we need to think about this case a bit more.

Reply:
Do you mean I need to bind this to Union instead of Primitive?

Reply:
exactly

Reply:
Done. See [the latest commit](https://github.com/flyteorg/flytekit/pull/2401/commits/c9cb9df622b8ee0746ccb7d68490eb6d22412526) for details. Here I added a bool argument `is_union_type_variants_expanded`. And if the default argument is used, then `is_union_type_variants_expanded` will be set to `False`, which will bypass the for loop expansion for union types here
https://github.com/flyteorg/flytekit/blob/c9cb9df622b8ee0746ccb7d68490eb6d22412526/flytekit/core/promise.py#L706-L718

Also, I added two tests for testing `Union[List[int]]`.